# Исследование надежности заемщиков

<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Цели,-план-и-методы-исследования-проекта-&quot;Исследование-надежности-заемщиков&quot;." data-toc-modified-id="Цели,-план-и-методы-исследования-проекта-&quot;Исследование-надежности-заемщиков&quot;.-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Цели, план и методы исследования проекта "Исследование надежности заемщиков".</a></span><ul class="toc-item"><li><span><a href="#Главные-вопросы-на-которые-предстоит-ответить-в-рамках-данного-ислледования,-явялются:" data-toc-modified-id="Главные-вопросы-на-которые-предстоит-ответить-в-рамках-данного-ислледования,-явялются:-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Главные вопросы на которые предстоит ответить в рамках данного ислледования, явялются:</a></span></li><li><span><a href="#План-исследования" data-toc-modified-id="План-исследования-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>План исследования</a></span><ul class="toc-item"><li><span><a href="#Проведем-предобработку-данных-датафрема,-в-том-числе:" data-toc-modified-id="Проведем-предобработку-данных-датафрема,-в-том-числе:-1.2.1"><span class="toc-item-num">1.2.1&nbsp;&nbsp;</span>Проведем предобработку данных датафрема, в том числе:</a></span></li><li><span><a href="#Во-второй-части-проведем-анализ-данных-обработанного-датафрема-для-получения-ответов-на-поставленные-в-целях-исследования-вопросы." data-toc-modified-id="Во-второй-части-проведем-анализ-данных-обработанного-датафрема-для-получения-ответов-на-поставленные-в-целях-исследования-вопросы.-1.2.2"><span class="toc-item-num">1.2.2&nbsp;&nbsp;</span>Во второй части проведем анализ данных обработанного датафрема для получения ответов на поставленные в целях исследования вопросы.</a></span></li><li><span><a href="#Подготовим-итоговые-выводы-по-проекту." data-toc-modified-id="Подготовим-итоговые-выводы-по-проекту.-1.2.3"><span class="toc-item-num">1.2.3&nbsp;&nbsp;</span>Подготовим итоговые выводы по проекту.</a></span></li></ul></li></ul></li><li><span><a href="#Загрузка-и-предобработка-данных" data-toc-modified-id="Загрузка-и-предобработка-данных-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Загрузка и предобработка данных</a></span><ul class="toc-item"><li><span><a href="#Предобработка-данных" data-toc-modified-id="Предобработка-данных-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Предобработка данных</a></span><ul class="toc-item"><li><span><a href="#Удаление-пропусков" data-toc-modified-id="Удаление-пропусков-2.1.1"><span class="toc-item-num">2.1.1&nbsp;&nbsp;</span>Удаление пропусков</a></span></li></ul></li><li><span><a href="#Обработка-аномальных-значений" data-toc-modified-id="Обработка-аномальных-значений-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Обработка аномальных значений</a></span></li><li><span><a href="#Удаление-пропусков-(продолжение)" data-toc-modified-id="Удаление-пропусков-(продолжение)-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Удаление пропусков (продолжение)</a></span></li><li><span><a href="#Изменение-типов-данных" data-toc-modified-id="Изменение-типов-данных-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Изменение типов данных</a></span></li><li><span><a href="#Обработка-дубликатов" data-toc-modified-id="Обработка-дубликатов-2.5"><span class="toc-item-num">2.5&nbsp;&nbsp;</span>Обработка дубликатов</a></span></li><li><span><a href="#Категоризация-данных" data-toc-modified-id="Категоризация-данных-2.6"><span class="toc-item-num">2.6&nbsp;&nbsp;</span>Категоризация данных</a></span></li></ul></li><li><span><a href="#Исселедваотельский-анализ-данных" data-toc-modified-id="Исселедваотельский-анализ-данных-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Исселедваотельский анализ данных</a></span><ul class="toc-item"><li><span><a href="#Есть-ли-зависимость-между-количеством-детей-и-возвратом-кредита-в-срок?" data-toc-modified-id="Есть-ли-зависимость-между-количеством-детей-и-возвратом-кредита-в-срок?-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Есть ли зависимость между количеством детей и возвратом кредита в срок?</a></span></li><li><span><a href="#Есть-ли-зависимость-между-семейным-положением-и-возвратом-кредита-в-срок?" data-toc-modified-id="Есть-ли-зависимость-между-семейным-положением-и-возвратом-кредита-в-срок?-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>Есть ли зависимость между семейным положением и возвратом кредита в срок?</a></span></li><li><span><a href="#Есть-ли-зависимость-между-уровнем-дохода-и-возвратом-кредита-в-срок?" data-toc-modified-id="Есть-ли-зависимость-между-уровнем-дохода-и-возвратом-кредита-в-срок?-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>Есть ли зависимость между уровнем дохода и возвратом кредита в срок?</a></span></li><li><span><a href="#Как-разные-цели-кредита-влияют-на-его-возврат-в-срок?" data-toc-modified-id="Как-разные-цели-кредита-влияют-на-его-возврат-в-срок?-3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span>Как разные цели кредита влияют на его возврат в срок?</a></span><ul class="toc-item"><li><span><a href="#Получение-образования---одна-из-групп-с-высоким-процентом-невозврата-в-срок,-рассмотрим-эту-группу-сгруппировав-получателей-займа-на-получение-образования-по-уровню-образования" data-toc-modified-id="Получение-образования---одна-из-групп-с-высоким-процентом-невозврата-в-срок,-рассмотрим-эту-группу-сгруппировав-получателей-займа-на-получение-образования-по-уровню-образования-3.4.1"><span class="toc-item-num">3.4.1&nbsp;&nbsp;</span>Получение образования - одна из групп с высоким процентом невозврата в срок, рассмотрим эту группу сгруппировав получателей займа на получение образования по уровню образования</a></span></li></ul></li><li><span><a href="#Приведите-возможные-причины-появления-пропусков-в-исходных-данных." data-toc-modified-id="Приведите-возможные-причины-появления-пропусков-в-исходных-данных.-3.5"><span class="toc-item-num">3.5&nbsp;&nbsp;</span>Приведите возможные причины появления пропусков в исходных данных.</a></span></li><li><span><a href="#Объясните,-почему-заполнить-пропуски-медианным-значением-—-лучшее-решение-для-количественных-переменных." data-toc-modified-id="Объясните,-почему-заполнить-пропуски-медианным-значением-—-лучшее-решение-для-количественных-переменных.-3.6"><span class="toc-item-num">3.6&nbsp;&nbsp;</span>Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.</a></span></li></ul></li><li><span><a href="#Общий-вывод" data-toc-modified-id="Общий-вывод-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Общий вывод</a></span><ul class="toc-item"><li><span><a href="#В-первой-части-проведена-предобработка-датафрема,-в-том-числе:" data-toc-modified-id="В-первой-части-проведена-предобработка-датафрема,-в-том-числе:-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>В первой части проведена предобработка датафрема, в том числе:</a></span></li><li><span><a href="#Во-второй-части-исследования-проведен-анализ-данных-на-предмет-оценки-следущих-вопросов:" data-toc-modified-id="Во-второй-части-исследования-проведен-анализ-данных-на-предмет-оценки-следущих-вопросов:-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>Во-второй части исследования проведен анализ данных на предмет оценки следущих вопросов:</a></span></li></ul></li></ul></div>

## Цели, план и методы исследования проекта "Исследование надежности заемщиков".

### Главные вопросы на которые предстоит ответить в рамках данного ислледования, явялются:

        1. Есть ли зависимость между количеством детей и возвратом кредита в срок?
        2. Есть ли зависимость между семейным положением и возвратом кредита в срок?
        3. Есть ли зависимость между уровнем дохода и возвратом кредита в срок?
        4. Как разные цели кредита влияют на его возврат в срок?
        5. Возможные причины появления пропусков в исходных данных.
        6. Почему заполнить пропуски медианным значением — лучшее решение для количественных переменных?

### План исследования

 #### Проведем предобработку данных датафрема, в том числе:
     - найдем пропуски значений и опеределим метод их заполнения, если это возможно;
     - выявим аномалии, экстремальные или нелогичные значения, выберем метод исправления или удаления, если исправить невозможно и показатели артефактов несущесвтенно влияют на данные;
     - проанализируем типы данных в столбцах и исправим, при необходимости;
     - проанализируем датафрейм на предмет наличия явных и неявных дубликатов: приведем значения в столбцах напиванные строчными и заглавными буквами к нижнему регистру, затем найдем все явные дубликаты и удалим их;
     - для дальнейшего анализа (вопросы 3,4) и группировки данных по уровню дохода и целям кредита, создадим две дополнительных колонки, в одной из которых распределим всех заемщиков на пять групп по уровню дохода, во-второй, унифицируем название целей кредита, исходя из представленной информации от заемщиков.
#### Во второй части проведем анализ данных обработанного датафрема для получения ответов на поставленные в целях исследования вопросы.
     - Для этого по вопросам 1-4 сформируем сгруппированные по указанным в вопросах критериям (дети, семейное положение, доход, цели займа) срезы данных и рассчитаем процент невозвратов в срок по каждой группе, входящей в срез. 
     - Проанализируем полученные результаты, при необходимостиб перегрупперуем данные для более корректного анализа и сформируем выводы по каждому вопросу.
 
#### Подготовим итоговые выводы по проекту. 

## Загрузка и предобработка данных

In [43]:
import pandas as pd

In [44]:
path = 'C:\\Users\\GAV\\Desktop\\Study\\BA\\PythonScripts\\Portfolio\\Total\\0_Итог\\Datasets\\data.csv'

data = pd.read_csv(path)

In [45]:
data.head()

children  days_employed  dob_years education  education_id  \
0         1   -8437.673028         42    высшее             0   
1         1   -4024.803754         36   среднее             1   
2         0   -5623.422610         33   Среднее             1   
3         3   -4124.747207         32   среднее             1   
4         0  340266.072047         53   среднее             1   

      family_status  family_status_id gender income_type  debt   total_income  \
0   женат / замужем                 0      F   сотрудник     0  253875.639453   
1   женат / замужем                 0      F   сотрудник     0  112080.014102   
2   женат / замужем                 0      M   сотрудник     0  145885.952297   
3   женат / замужем                 0      M   сотрудник     0  267628.550329   
4  гражданский брак                 1      F   пенсионер     0  158616.077870   

                      purpose  
0               покупка жилья  
1     приобретение автомобиля  
2               покупка жилья  
3  дополнительное образование  
4             сыграть свадьбу

In [46]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


### Предобработка данных

#### Удаление пропусков

In [47]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [48]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [49]:
data['days_employed'] = data['days_employed'].abs()

**Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [50]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть.

**Выведите перечень уникальных значений столбца `children`.**

In [51]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5], dtype=int64)

**В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [52]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [53]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5], dtype=int64)

### Удаление пропусков (продолжение)

**Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [54]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [55]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [56]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру.**

In [57]:
data['education'] = data['education'].str.lower()

**Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [58]:
data.duplicated().sum()

71

In [59]:
data = data.drop_duplicates()

### Категоризация данных

**На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [60]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E - до 0,03 млн. руб.'
        elif 30001 <= income <= 50000:
            return 'D- от 0,03 до 0,05 млн. руб.'
        elif 50001 <= income <= 200000:
            return 'C- от 0,05 до 0,2 млн. руб.'
        elif 200001 <= income <= 1000000:
            return 'B - от 0,2 до 1 млн. руб.'
        elif income >= 1000001:
            return 'A - больше 1 млн. руб.'
    except:
        pass

In [61]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [62]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [63]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [64]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

## Исселедваотельский анализ данных

### Есть ли зависимость между количеством детей и возвратом кредита в срок?

In [65]:
#создадим функцию для группировки срезов данных по необходимым для последующего анализа критериям (дети, семейное положение, доход, цели кредита)
def group_and_agg(dataframe, group_by_column, aggregation_column):
    # расчитаем общее количество заемщиков по каждой группе и количество невозвратов по ним же.
    grouped = dataframe.groupby(group_by_column).agg({aggregation_column: ['count', 'sum']})
    # рассчитаем долю каждой группы в общем количестве заемщиков в колонке 'struct_%'
    grouped['struct_%'] = (grouped['debt']['count'] /grouped['debt']['count'].sum()).multiply(100)
    # расчитаем в столбце 'non_repayment_%' процент тех, кто имел задолженность по возврату кредитов для каждой групппы
    grouped['non_repayment_%'] = (grouped['debt']['sum'] /grouped['debt']['count']).multiply(100)
    return grouped

In [66]:
# Вызов функции группировки по критерию 'children'
child_grouped = group_and_agg(data, 'children', 'debt')
child_grouped

debt         struct_% non_repayment_%
          count   sum                           
children                                        
0         14091  1063  66.058788        7.543822
1          4808   444  22.539965        9.234609
2          2052   194   9.619802        9.454191
3           330    27   1.547044        8.181818
4            41     4   0.192209        9.756098
5             9     0   0.042192        0.000000

In [67]:
# переименуем столбы для наглядности
child_grouped = child_grouped.rename(columns={'debt':'Заемщики','count':'Кол-во заемщиков', 'sum':'Кол-во просрочек','struct_%':'Структура заемщиков по группам,%', 'non_repayment_%':'Доля неплательщиков,%'})
child_grouped

Заемщики                  Структура заемщиков по группам,%  \
         Кол-во заемщиков Кол-во просрочек                                    
children                                                                      
0                   14091             1063                        66.058788   
1                    4808              444                        22.539965   
2                    2052              194                         9.619802   
3                     330               27                         1.547044   
4                      41                4                         0.192209   
5                       9                0                         0.042192   

         Доля неплательщиков,%  
                                
children                        
0                     7.543822  
1                     9.234609  
2                     9.454191  
3                     8.181818  
4                     9.756098  
5                     0.000000

**Вывод:** 

1. Для ответа на вопрос, есть ли зависимость между количеством детей и возвратом кредита в срок датафрейм был сгруппирован по количеству детей и рассчитан процен невозврата задолженности в срок по каждой группе: с детьми и без детей.

2. Наиболее существеную долю заемщиков с детьми занимают группы с 1-2 детьми, доля займов с 3-5 детьми менее 5%, поэтому для анализа наиболее важны показатели групп 1-2 ребенком.

3. Исходя из представленной таблицы процент невозврата в срок заемщиков без детей составляет 7,54%, а процент невозврата в срок с 1-2 детьми возрастает до 9,2% - 9,45% (выше минимум на 22%, чем без детей)

  4. Таким образом, есть прямая зависимость, что заемщики с детьми чаще не возвращают заем в срок, чем заемщики без детей. 
    При этом, разброс между заемщиками с 1-м ребенком и 2-мя детьми 9,2% против 9,45%, всего 0,25%. То есть, с увеличеним количества детей, не выявлено существеннго увеличения процента невозврата задолженности в срок.

### Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [68]:
# Вызов функции группировки по критерию 'family_status'
family_grouped = group_and_agg(data, 'family_status', 'debt')
family_grouped

debt        struct_% non_repayment_%
                       count  sum                           
family_status                                               
Не женат / не замужем   2796  273  13.107684        9.763948
в разводе               1189   84   5.574047        7.064760
вдовец / вдова           951   63   4.458300        6.624606
гражданский брак        4134  385  19.380245        9.313014
женат / замужем        12261  927  57.479724        7.560558

Для дальнейшего анализа объединим группы "в разводе" и "вдовец/вдова" с "не женат/не замужем", по следующим причинам:
    1. фактически это люди ведущие свое домашнее хозяйство в одиночку; 
    2. стутс чкловека "не женат/не замужем" не исключают того, что этот человек разведен/вдовец/вдова;
    3. доли этих групп в общей структуре заемщиков наименьшие.
Группу "гражданский брак" оставим без объединения с группой "женат/замужем", так как группа занимает существенную долю в структуре заемщиков и процент невозврата в срок кредитов существенно отличается от группы "женат/замужем".

In [69]:
#объединим группы "в разводе" и "вдовец/вдова" с "не женат/не замужем", 
family_status_new = data 
family_status_new.loc[(family_status_new['family_status'] == 'в разводе') | (family_status_new['family_status'] == 'вдовец / вдова'), 'family_status'] = 'Не женат / не замужем' 
family_status_new['family_status'].unique()

array(['женат / замужем', 'гражданский брак', 'Не женат / не замужем'],
      dtype=object)

In [70]:
# Вызов функции группировки по критерию 'family_status'
family_grouped = group_and_agg(data, 'family_status', 'debt')
family_grouped

debt        struct_% non_repayment_%
                       count  sum                           
family_status                                               
Не женат / не замужем   4936  420  23.140031        8.508914
гражданский брак        4134  385  19.380245        9.313014
женат / замужем        12261  927  57.479724        7.560558

In [71]:
# переименуем столбы для наглядности
family_grouped = family_grouped.rename(columns={'debt':'Заемщики','count':'Кол-во заемщиков', 'sum':'Кол-во просрочек','struct_%':'Структура заемщиков по группам,%', 'non_repayment_%':'Доля неплательщиков,%'})
family_grouped

Заемщики                   \
                      Кол-во заемщиков Кол-во просрочек   
family_status                                             
Не женат / не замужем             4936              420   
гражданский брак                  4134              385   
женат / замужем                  12261              927   

                      Структура заемщиков по группам,% Доля неплательщиков,%  
                                                                              
family_status                                                                 
Не женат / не замужем                        23.140031              8.508914  
гражданский брак                             19.380245              9.313014  
женат / замужем                              57.479724              7.560558

**Вывод:** 

1. Для ответа на вопрос, есть ли зависимость между семейным положением и возвратом кредита в срок датафрейм был сгруппирован по семейному статусу и рассчитан процен невозврата задолженности в срок по каждой группе.

2. В рамках анализа две группы "в разводе" и "вдовец/вдова" были объединены "не женат/не замужем", так как: 
    1. фактически это люди ведущие свое домашнее хозяйство в одиночку; 
    2. стутс чкловека "не женат/не замужем" не исключают того, что этот человек разведен/вдовец/вдова;
    3. доли этих групп в общей структуре заемщиков наименьшие.

В итоге анализировали 3 группы: 'женат / замужем', 'гражданский брак', 'Не женат / не замужем'

3. Исходя из анализа итоговой таблицы:
    1. наименьший процент невозврата в срок заемщиков из группы 'женат / замужем'- 7,56%, 
    2. процент невозврата в срок заемщиков из группы 'Не женат / не замужем' - 8,5%, 
    3. процент невозврата в срок заемщиков из группы 'гражданский брак' самый высокий - 9,3%

    4.Таким образом, есть прямая зависимость, между семейным положением и возвратом кредита в срок.
     Наименьший процент невозврата в срок у заемщиков официально зарегистрированных в браке - 7,56%, далее идет группа не женатых/не замужних людей, процент невозврата в срок которых - 8,5%, что выше на 12,4% от группы людей с зарегистрированным в браком и наибольщая доля невозварата в срок у заемщиков в гражданском браке - 9,3%, выше на 23% от группы людей с зарегистрированным в браком).

### Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [72]:
# Вызов функции группировки по критерию 'total_income_category'
icome_grouped = group_and_agg(data, 'total_income_category', 'debt')
icome_grouped

debt         struct_% non_repayment_%
                              count   sum                           
total_income_category                                               
A - больше 1 млн. руб.           25     2   0.117200        8.000000
B - от 0,2 до 1 млн. руб.      5014   354  23.505696        7.060231
C- от 0,05 до 0,2 млн. руб.   15921  1353  74.637851        8.498210
D- от 0,03 до 0,05 млн. руб.    349    21   1.636116        6.017192
E - до 0,03 млн. руб.            22     2   0.103136        9.090909

1. Наиболее сущесвтенная доля заемшиков в группах 'B' и 'C', они занимают порядка 98% от всех заемщиков, остальные группы все вместе составляют мнее 2% от все базы, анализировать их отдельно невозможно, так как выборка слишком мала, можно сделать необоснованные выводы.
2. Объединим группы 'D' и 'E'  с группоу 'C', а группу 'А' с группой 'B' для дальнейшего анализа, сформируем две группы: группа 'А' с доходом от 0,2 млн. рую и выше и группу 'B' с доходом до 0,2 млн. руб.

In [73]:
def categorize_income_new(income):
    try:
        if 0 <= income <= 200000:
            return 'B - до 0,2 млн. руб.'
        elif income >= 200001:
            return 'A - от 0,2 млн. руб. и выше'
    except:
        pass

In [74]:
data['income_category_group'] = data['total_income'].apply(categorize_income_new)

In [75]:
# Вызов функции группировки по критерию 'income_category_group'
icome_grouped = group_and_agg(data, 'income_category_group', 'debt')
icome_grouped

debt         struct_% non_repayment_%
                             count   sum                           
income_category_group                                              
A - от 0,2 млн. руб. и выше   5039   356  23.622896        7.064894
B - до 0,2 млн. руб.         16292  1376  76.377104        8.445863

In [76]:
# переименуем столбы для наглядности
icome_grouped = icome_grouped.rename(columns={'debt':'Заемщики','count':'Кол-во заемщиков', 'sum':'Кол-во просрочек','struct_%':'Структура заемщиков по группам,%', 'non_repayment_%':'Доля неплательщиков,%'})
icome_grouped

Заемщики                   \
                            Кол-во заемщиков Кол-во просрочек   
income_category_group                                           
A - от 0,2 млн. руб. и выше             5039              356   
B - до 0,2 млн. руб.                   16292             1376   

                            Структура заемщиков по группам,%  \
                                                               
income_category_group                                          
A - от 0,2 млн. руб. и выше                        23.622896   
B - до 0,2 млн. руб.                               76.377104   

                            Доля неплательщиков,%  
                                                   
income_category_group                              
A - от 0,2 млн. руб. и выше              7.064894  
B - до 0,2 млн. руб.                     8.445863

**Вывод:** 

1. Для ответа на вопрос, есть ли зависимость между уровнем дохода и возвратом кредита в срок датафрейм был сгруппирован по группам дохода и рассчитан процент невозврата задолженности в срок по каждой группе.

2. В рамках анализа было выявлено, что две группы "B - от 0,2 до 1 млн. руб." и "C- от 0,05 до 0,2 млн. руб." занимают в структуре порядка 98% от всего количесва заемщиков, остальные группы в связи с их малочисленностью были объединены следующим образом: группы 'D' и 'E'  с группоу 'C', а группу 'А' с группой 'B' и для анализа, сформированы две группы: группа 'А' с доходом от 0,2 млн. рую и выше и группу 'B' с доходом до 0,2 млн. руб.

  3. В результате выявлена следующая зависмость у группы с доходом от 0,2 млн. руб и выше процент невозварата задолженности в срок составил 7,06%, а у группы с доходом до 0,2 млн. руб. 8,45%, что на 19,6% выше, то есть чем выше доход, тем процент невозварата займа в срок меньше.

### Как разные цели кредита влияют на его возврат в срок?

In [77]:
# Вызов функции группировки по критерию 'purpose_category'
goals_grouped = group_and_agg(data, 'purpose_category', 'debt')
goals_grouped

debt        struct_% non_repayment_%
                          count  sum                           
purpose_category                                               
операции с автомобилем     4279  400  20.060007        9.347978
операции с недвижимостью  10751  780  50.400825        7.255139
получение образования      3988  369  18.695795        9.252758
проведение свадьбы         2313  183  10.843373        7.911803

#### Получение образования - одна из групп с высоким процентом невозврата в срок, рассмотрим эту группу сгруппировав получателей займа на получение образования по уровню образования

In [78]:
#выгрузим данные по заемщикам с целью займа 'получение образования' 
education_groupp = data.loc[data['purpose_category'] == 'получение образования']
# Вызов функции группировки по критерию 'education'
education_groupp = group_and_agg(education_groupp, 'education', 'debt')
education_groupp

debt        struct_% non_repayment_%
                    count  sum                           
education                                                
высшее                928   55  23.269809        5.926724
начальное              55    5   1.379137        9.090909
неоконченное высшее   131   12   3.284855        9.160305
среднее              2873  297  72.041123       10.337626
ученая степень          1    0   0.025075        0.000000

Из таблицы выше видно, что наибольший процент задержек по замам у заемщиков без высшего образования от 9% до 10,33%.

In [79]:
# переименуем столбы для наглядности
goals_grouped = goals_grouped.rename(columns={'debt':'Заемщики','count':'Кол-во заемщиков', 'sum':'Кол-во просрочек','struct_%':'Структура заемщиков по группам,%', 'non_repayment_%':'Доля неплательщиков,%'})
goals_grouped

Заемщики                   \
                         Кол-во заемщиков Кол-во просрочек   
purpose_category                                             
операции с автомобилем               4279              400   
операции с недвижимостью            10751              780   
получение образования                3988              369   
проведение свадьбы                   2313              183   

                         Структура заемщиков по группам,%  \
                                                            
purpose_category                                            
операции с автомобилем                          20.060007   
операции с недвижимостью                        50.400825   
получение образования                           18.695795   
проведение свадьбы                              10.843373   

                         Доля неплательщиков,%  
                                                
purpose_category                                
операции с автомобилем                9.347978  
операции с недвижимостью              7.255139  
получение образования                 9.252758  
проведение свадьбы                    7.911803

**Вывод:** 

1. Наименьший процент невозвратов в срок по займам для операций с недвижимостью 7,25%. Скорее всего это объясняется тем, что заем выдается под залог невижимости. Для подтверждения необходима дополнительная информация по условиям предоставляемых займов для операций с недвижимостью.

2. Вторая цель с наименьшим процентом невозврата в срок - это проведение свадьбы - 7,91%. Этот результат перекликается с анализом возврата кредита в срок в зависмиости от семейного положения, в той части анализ показал 7,56% неаозвратов в срок по группе 'жента/замужем'.

3. Наиболее рискованные цели с точки зрения своевременности возврата займа в срок:
- на получение образования (9,25%). На получение образования в большинстве своем берут займ заемщики  без высшего образования, со средним и неоконченным высшим, что увеличивает риск задержки по выплате займа в срок, так по групппе заемшиков со средним образованием невыплаты в срок достигают 10,3% 
- на операции с автомобилем (9,34%).

### Приведите возможные причины появления пропусков в исходных данных.

*Ответ:* Причинами появления пропусков в исходных данных могут быть как:
1. Технологические ошибки, которые возникают при копировании, считывании, смене формата, записи данных. 
2. Пропуски связанные с человеческим фактором, если данные вносятся вручную и/или нет четких инструкций по заполнению формата данных.

В анализируемом датафрейме пропуски были обнаружены в двух колонках 'days_employed' и 'total_income' в одинаковом количестве 2174, так как у двух разных колонок одинаковое количество пропусков, то скорее всего проблема в выгрузке данных, а не в их ручном заполнеии заполнении.

### Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:* Медиана  —  это такое число в выборке, что ровно половина элементов больше него, а другая половина —  меньше. Применение медианы позволяет исключить влияние на средний показатель экстремальных значений в выборке, в отличие, например,
от среднего значения. Пэтому, если медиана применяется к однородной выборке, элементы котрой обладают одинаковыми признаками, то применение медианы позволит наиболее точно заполнить пропуски количественных переменных.

## Общий вывод

Проведен анализ представленной базы данных на предмет исследования надежности заемщиков. Работа была проведена в два этапа.

### В первой части проведена предобработка датафрема, в том числе:
    - найдены пропуски значений в колонках 'days_employed' и 'total_income' в колчестве 2174 шт., это  порядка 10% от общей базы, что существенно, поэтому эти пропуски были заполнены медианными значениями как наиболее подходящие показатели для таких количественных переменных как 'days_employed' и 'total_income', при этом медиана была рссчитана по группам, отдельно для каждого типа занятости/('income_type');
    - одинаковое количество пропусков 2174 в двух колонках с таким разными данными как 'days_employed' и 'total_income', скорее всего говорит о возможном техническом сбое при выгрузке/загрузке данных; 
     - выявленные аномалии с данными о количестве детей были удалены, т.к. не были сущесвтенными для датафрейма;
     - выявлена аномалия в медианном количестве дней_работы/'days_employed' по типам_занятости/('income_type') - пенсионеры/безработныне, аномалия не корректировалась, т.к. показатели не участвовали в данном анализе;
     - тип данных в столбцах 'total_income' приведен к целочисленному значению для более наглядного представления данных о доходах;
     - устранены явные и нефвные дубликаты в данных;
     - подготовлена в отдельной колонке дополнительная информация для группировки заемщиков по 5 группам доходов, а также в отдельной колонке унифицированы названия целей кредита, исходя из представленной информации от заемщиков.


### Во-второй части исследования проведен анализ данных на предмет оценки следущих вопросов:

    • Есть ли зависимость между количеством детей и возвратом кредита в срок? 
 
     Ответ: 
Да, выявлена зависимость, в том что заемщики с детьми чаще не возвращают заем в срок, чем заемщики без детей. При этом, разброс между заемщиками с 1-м ребенком и 2-мя детьми 9,2% против 9,45%, всего 0,25%. То есть, с увеличеним количества детей, не выявлено существеннго увеличения процента невозврата задолженности в срок.

     • Есть ли зависимость между семейным положением и возвратом кредита в срок?
     Ответ:
Да, выявлена зависимость, между семейным положением и возвратом кредита в срок.
Наименьший процент невозврата в срок у заемщиков официально зарегистрированных в браке - 7,56%, далее идет группа не женатых/не замужних заемщиков, процент невозврата в срок которых - 8,5%, что выше на 12,4% от группы людей с зарегистрированным в браком и наибольщая доля невозварата в срок у заемщиков в гражданском браке - 9,3%, выше на 23% от группы людей с зарегистрированным в браком.

    •	Есть ли зависимость между уровнем дохода и возвратом кредита в срок?
    Ответ:
В результате выявлена следующая зависмость у группы с доходом от 0,2 млн. руб и выше процент невозварата задолженности в срок составил 7,06%, а у группы с доходом до 0,2 млн. руб. 8,45%, что на 19,6% выше, то есть чем выше доход, тем доля невозваратов займов в срок снижается.

    •  Как разные цели кредита влияют на его возврат в срок?
  Ответ:
1. Наименьший процент невозвратов в срок по займам для операций с недвижимостью 7,25%.
2. Вторая цель с наименьшим процентом невозврата в срок - это проведение свадьбы - 7,91%. Этот результат составим с уровнем с неаозвратов в срок по группе 'жента/замужем' - 7,56%, выявленном при анализе возвратов займов в срок в зависмиости от семейного положения.
3. Наиболее рискованные цели с точки зрения своевременности возврата займа в срок:
    - на получение образования (9,25%);
    - на операции с автомобилем (9,34%).